File: src/main/java/macro303/neptunes/Star.java
package macro303.neptunes;

import com.google.gson.annotations.SerializedName;
import org.jetbrains.annotations.NotNull;

/**
 * Created by Macro303 on 2018-05-07.
 */
public class Star {
	@NotNull
	@SerializedName("uid")
	private Integer uid = -1;           //Star ID
	@NotNull
	@SerializedName("n")
	private String n = "Unknown";       //Name
	@NotNull
	@SerializedName("puid")
	private Integer puid = -1;          //Player ID
	@NotNull
	@SerializedName("v")
	private String v = "Unknown";       //Visible
	@NotNull
	@SerializedName("y")
	private Double y = -1.0;            //Y Coordinate
	@NotNull
	@SerializedName("x")
	private Double x = -1.0;            //X Coordinate

	@Override
	public boolean equals(Object o) {
		if (this == o) return true;
		if (!(o instanceof Star)) return false;

		Star star = (Star) o;

		if (!uid.equals(star.uid)) return false;
		if (!n.equals(star.n)) return false;
		if (!puid.equals(star.puid)) return false;
		if (!v.equals(star.v)) return false;
		if (!y.equals(star.y)) return false;
		return x.equals(star.x);
	}

	@Override
	public int hashCode() {
		int result = uid.hashCode();
		result = 31 * result + n.hashCode();
		result = 31 * result + puid.hashCode();
		result = 31 * result + v.hashCode();
		result = 31 * result + y.hashCode();
		result = 31 * result + x.hashCode();
		return result;
	}

	@Override
	public String toString() {
		return "Star{" +
				"uid=" + uid +
				", n='" + n + '\'' +
				", puid=" + puid +
				", v='" + v + '\'' +
				", y=" + y +
				", x=" + x +
				'}';
	}
}

File: src/main/kotlin/macro303/neptunes_pride/star/Star.kt
package macro303.neptunes_pride.star

import com.google.gson.annotations.SerializedName

internal data class Star(
	@SerializedName(value = "uid") val starID: Int,
	@SerializedName(value = "n") val name: String,
	@SerializedName(value = "puid") val playerID: Int,
	private val v: String,
	@SerializedName(value = "y") val yCoordinate: Double,
	@SerializedName(value = "x") val xCoordinate: Double
) : Comparable<Star> {
	val visible: Boolean
		get() = v == "1"

	override fun compareTo(other: Star): Int {
		return compareBy<Star>(
			{ it.playerID },
			{ it.name }
		).reversed().compare(this, other)
	}

	override fun toString(): String {
		return "Star(starID=$starID, name='$name', playerID=$playerID, v='$v', yCoordinate=$yCoordinate, xCoordinate=$xCoordinate)"
	}
}